### PR TITLE
[skip ci] Update TextCNN example to torchtext==0.9.0 api

### LINKED ISSUE #1900: Replace legacy code in `TextCNN` example
## 🚀 Feature
`torchtext` in their latest release `0.9.0` moved a lot of features to legacy code, and because of that we temporary updated our [TextCNN](https://github.com/pytorch/ignite/blob/master/examples/notebooks/TextCNN.ipynb) example to compatible with the legacy code in #1890 (as quick fix for now).
Now we need to replace the legacy code in this example with the provided features in `torchtext==0.9.0`.
- `torchtext==0.9.0` release note: https://github.com/pytorch/text/releases/tag/v0.9.0-rc5
- Migration guide from `torchtext` to show how to replace legacy code: https://github.com/pytorch/text/blob/master/examples/legacy_tutorial/migration_tutorial.ipynb

### PR DESCRIPTION
**Pull request by cozek**
Fixes #1900

Description:
Remove legacy torchtext code from TextCNN example notebook and update to torchtext==0.9.0 API

Check list:
- [x] Get notebook running
- [x] Test in CPU/GPU in Colab

### CODE REVIEW

**Comment by cozek** [comment]
Some comments:

1. I changed from `en` to `en_core_web_sm` since `en` is deprecated.
`python -m spacy download en_core_web_sm`

2. Set the seed for random using `random.seed(SEED)`
3. Set the tokenizer to spacy since `spacy` was never used in the original notebook even though installed. Perhaps it is used internally.
     `tokenizer = get_tokenizer("spacy")`
4. Used only 1000 samples to speed up testing and provided option for using the entire set. Since, `torchtext` no longer handles custom splitting AFAIK.
```python
# We are using only 1000 samples for faster training
# set to -1 to use full data
N = 1000 

# We will use 80% of the `train split` for training and the rest for validation
train_frac = 0.8
_temp = list(train_iter)


random.shuffle(_temp)
_temp = _temp[:(N if N > 0 else len(_temp) )]
n_train = int(len(_temp)*train_frac)

train_list = _temp[:n_train]
validation_list = _temp[n_train:]
test_list = list(test_iter)
test_list = test_list[:(N if N > 0 else len(test_list))]
```
5. I couldn't get bucket iterator running as described in the migration guide, so I skipped it. I followed the migration guide as closely as I could.
6. `process_function` and `eval_function` had to be changed.

**Review by KickItLikeShika — APPROVED** [review]
Thanks for the PR @cozek! LGTM!

**Comment by KickItLikeShika** [comment]
@cozek i think the bucket iterator part is not a must, and the changes in `process_function` and `eval_function` is a must because of the device issues and that's discussed before in the issue. Regarding the the size of training data you can wait for @sdesrozis or @vfdev-5 reviews.

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,9 @@ coverage.xml
 /docs/src/
 /docs/build/
 /docs/source/generated/
+
+# Jupyter
+examples/notebooks/.ipynb_checkpoints/
+
+# System files
+.DS_Store
```
Comment (vfdev-5) [inline .gitignore:47]: Let's keep this file as it was, please.
You can ignore `.ipynb_checkpoints` and `.DS_Store` locally with `.git/info/exclude` ... We haven't yet decided on if we should exclude these files repository-wise...

**Review by vfdev-5 — COMMENTED** [review]
@cozek thanks for the PR, I'll review it in details later. I have already a comment on .gitignore modifications and also wonder if we should recompute vocabulary this way : 
```python


from collections import Counter
from torchtext.vocab import Vocab

counter = Counter()

for (label, line) in train_list:
    counter.update(tokenizer(line))

vocab = Vocab(
    counter,
    min_freq=10,
    vectors=GloVe(name='6B', dim=100, cache='/tmp/glove/')
)

```

Can't we reuse a predefined vocabulary or make sequential counter update into a parallel to make it a bit faster ?

**Comment by cozek** [comment]
@vfdev-5  I am simply trying to emulate 
`TEXT.build_vocab(train_data, vectors=GloVe(name='6B', dim=100, cache='/tmp/glove/'))`

AFAIK GloVe is predefined in `torchtext.vocab`. I don't think I fully understand what you mean. 
If you show me an example of what needs to be changed to make it faster I am happy to incorporate it. :)
Thanks!

File: examples/notebooks/TextCNN.ipynb [hunk 2]
```diff
@@ -1,853 +1,1016 @@
...
+  "metadata": {
+    "colab": {
+      "name": "Copy of TextCNN PR.ipynb",
+      "private_outputs": true,
+      "provenance": [],
+      "collapsed_sections": []
+    },
+    "kernelspec": {
+      "name": "python388jvsc74a57bd0bb24fb798fa891713af3d36fbae541dd86145d8cb277c7e680316fd96a4b69ba",
+      "display_name": "Python 3.8.8 64-bit ('ingite': conda)"
```
Comment (vfdev-5) [inline examples/notebooks/TextCNN.ipynb:13]: Could you please remove this metadata.

File: examples/notebooks/TextCNN.ipynb [hunk 3]
```diff
@@ -1,853 +1,1016 @@
...
+      },
+      "source": [
+        "import torch\n",
+        "import torch.nn as nn\n",
+        "import torch.nn.functional as F\n",
+        "\n",
+        "SEED = 1234\n",
+        "random.seed(SEED)\n",
+        "torch.manual_seed(SEED)\n",
+        "torch.cuda.manual_seed(SEED)"
```
Comment (vfdev-5) [inline examples/notebooks/TextCNN.ipynb:157]: Let's update here as well. We can use `ignite.utils.manual_seed(SEED)` instead of these 3 lines.

File: examples/notebooks/TextCNN.ipynb [hunk 4]
```diff
@@ -1,853 +1,1016 @@
...
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "qrlLB7PxkIW_"
+      },
+      "source": [
+        "random_sample = random.sample(train_list,1)[0]\n",
+        "print(' text:',random_sample[0])\n",
```
Comment (vfdev-5) [inline examples/notebooks/TextCNN.ipynb:327]:
```suggestion
        "print(' text:', random_sample[1])\n",
```

File: examples/notebooks/TextCNN.ipynb [hunk 5]
```diff
@@ -1,853 +1,1016 @@
...
+        "print('label:', random_sample[1])"
```
Comment (vfdev-5) [inline examples/notebooks/TextCNN.ipynb:328]:
```suggestion
        "print('label:', random_sample[0])"
```

File: examples/notebooks/TextCNN.ipynb [hunk 6]
```diff
@@ -1,853 +1,1016 @@
...
+        "def eval_function(engine, batch):\n",
+        "    model.eval()\n",
+        "    with torch.no_grad():\n",
+        "        y, x = batch\n",
+        "        y = y.to(device)\n",
+        "        x = x.to(device)\n",
+        "        y = y.float()\n",
+        "        y_pred = model(x)\n",
+        "        print('y_pred',y_pred)\n",
+        "        print('y',y)\n",
```
Comment (vfdev-5) [inline examples/notebooks/TextCNN.ipynb:753]: Let's remove these prints

**Review by vfdev-5 — COMMENTED** [review]
Sorry delayed review @cozek ! I added some comments to make it better. Could you please address them and it will be good to go. Thanks!

**Comment by vfdev-5** [comment]
> @vfdev-5 I am simply trying to emulate
> `TEXT.build_vocab(train_data, vectors=GloVe(name='6B', dim=100, cache='/tmp/glove/'))`
> 
> AFAIK GloVe is predefined in `torchtext.vocab`. I don't think I fully understand what you mean.
> If you show me an example of what needs to be changed to make it faster I am happy to incorporate it. :)
> Thanks!

I was thinking that there could be a way to build Vocab without recounting train set. However, it seems like even torchtext does that in their tutorials: https://pytorch.org/tutorials/beginner/text_sentiment_ngrams_tutorial.html
ok let's keep it like that here as well

**Comment by vfdev-5** [comment]
@KickItLikeShika could you please review this PR more thouroughly and check if all requested changed were applied ? Thanks

File: examples/notebooks/TextCNN.ipynb [hunk 7]
```diff
@@ -92,7 +101,7 @@
    },
    "outputs": [],
    "source": [
-    "from torchtext import data, legacy\n",
+    "from torchtext import data\n",
```
Comment (KickItLikeShika) [inline examples/notebooks/TextCNN.ipynb:104]: As we are not using `Fields` or `Labels`, please edit the MD part for the data.

File: examples/notebooks/TextCNN.ipynb [hunk 8]
```diff
@@ -168,149 +188,278 @@
...
-    "train_iterator, valid_iterator, test_iterator = legacy.data.BucketIterator.splits((train_data, valid_data, test_data), \n",
-    "                                                                           batch_size=32,\n",
-    "                                                                           device=device)"
+    "batch_size = 8  # A batch size of 8\n",
+    "\n",
+    "def create_iterators(batch_size=8):\n",
+    "    \"\"\"Heler function to create the iterators\"\"\"\n",
+    "    dataloaders = []\n",
+    "    for split in [train_list, validation_list, test_list]:\n",
+    "        bucket_dataloader = DataLoader(\n",
```
Comment (KickItLikeShika) [inline examples/notebooks/TextCNN.ipynb:435]: We are not creating `bucket_dataloaders` anymore, as we don't use a `batch_sampler`, please edit this variable name.

**Review by KickItLikeShika — COMMENTED** [review]
Thanks for the updates @cozek, you're doing a good job! Please just consider the reviews and we are good!

**Comment by KickItLikeShika** [comment]
@vfdev-5 I think everything is fine now with this PR

**Comment by vfdev-5** [comment]
Thanks a lot @KickItLikeShika for checking and @cozek for the update ! Let's merge it !
